contracts/shapella-upgrade/ShapellaUpgradeTemplate.sol

```diff
@@ -0,0 +1,870 @@
+        _assertContractVersion(_accountingOracle, EXPECTED_FINAL_ACCOUNTING_ORACLE_VERSION);
+        _assertContractVersion(_stakingRouter, EXPECTED_FINAL_STAKING_ROUTER_VERSION);
+        _assertContractVersion(_validatorsExitBusOracle, EXPECTED_FINAL_VALIDATORS_EXIT_BUS_ORACLE_VERSION);
+        _assertContractVersion(_withdrawalQueue, EXPECTED_FINAL_WITHDRAWAL_QUEUE_VERSION);
+        _assertContractVersion(_withdrawalVault, EXPECTED_FINAL_WITHDRAWAL_VAULT_VERSION);
+    }
+
+    function _assertContractVersion(IVersioned versioned, uint256 expectedVersion) internal view {
+        if (versioned.getContractVersion() != expectedVersion) {
+            revert InvalidContractVersion(address(versioned), expectedVersion);
+        }
+    }
+
+    function _transferOZAdminFromThisToAgent(IAccessControlEnumerable accessControlled) internal {
+        accessControlled.grantRole(DEFAULT_ADMIN_ROLE, _agent);
+        accessControlled.renounceRole(DEFAULT_ADMIN_ROLE, address(this));
+    }
+
+    function _resumeWithdrawalQueue() internal {
+        IWithdrawalQueue wq = _withdrawalQueue;
+        bytes32 resume_role = wq.RESUME_ROLE();
+        wq.grantRole(resume_role, address(this));
+        wq.resume();
+        wq.renounceRole(resume_role, address(this));
+    }
+
+    // To be strict need two almost identical _resume... function because RESUME_ROLE and resume()
+    // do not actually belong to PausableUntil contract
+    function _resumeValidatorsExitBusOracle() internal {
+        IValidatorsExitBusOracle vebo = _validatorsExitBusOracle;
+        bytes32 resume_role = vebo.RESUME_ROLE();
+        vebo.grantRole(resume_role, address(this));
+        vebo.resume();
+        vebo.renounceRole(resume_role, address(this));
+    }
+
+    error OnlyVotingCanUpgrade();
+    error CanOnlyStartOnce();
+    error CanOnlyFinishOnce();
+    error StartMustBeCalledBeforeFinish();
+    error UpgradeNotEnacted();
+    error LidoOracleMustNotBeUpgradedToLegacyYet();
+    error LidoOracleMustBeUpgradedToLegacy();
+    error WrongDsmOwner();
+    error WrongProxyAdmin(address proxy);
+    error WrongInitialImplementation(address proxy);
+    error InvalidContractVersion(address contractAddress, uint256 actualVersion);
+    error WrongOZAccessControlAdmin(address contractAddress);
+    error WrongOZAccessControlRoleHolders(address contractAddress, bytes32 role);
+    error NonZeroRoleHolders(address contractAddress, bytes32 role);
+    error WQNotResumed();
+    error VEBONotResumed();
+    error IncorrectOracleAndHashConsensusBinding(address oracle, address hashConsensus);
+    error IncorrectDepositSecurityModuleParameters(address _depositSecurityModule);
+    error WrongStakingModulesCount();
+    error InvalidOracleReportSanityCheckerConfig();
+    error WrongSealGateSealables();
+    error WrongStakingModuleParameters();
+    error WrongOracleDaemonConfigKeyValue(string key);
+    error WrongLocatorAddresses();
+}
```

Comment:
we were thinking yesterday about 'expirability' of the template (suppose a vote passed and failed to enact for some reason, it would be much better to disarm the template eventually).